### PR TITLE
reimplement self-reference replacement as a stylis plugin

### PR DESCRIPTION
**Pull request by probablyup**
this fixes the failing test case and another edge case

Fixes #2108 
Fixes the second issue noted toward the end of #2095

### CODE REVIEW

**Comment by codeaid** [comment]
Just adding a link to the test case I created to demonstrate the issue in #2095:
https://stackblitz.com/edit/react-9lm5xr

File: src/test/styles.test.js [hunk 1]
```diff
@@ -55,6 +55,10 @@ describe('with styles', () => {
         & + & {
           margin-bottom: 4px;
         }
+
+        & > & {
```
Comment (kitten) [inline src/test/styles.test.js]: After looking at the tests, do we have snapshots for these three combinators in particular?

Comment (probablyup) [reply src/test/styles.test.js]: Other than the ones in that group?

Comment (kitten) [reply src/test/styles.test.js]: sorry, my bad. I meant stylis level tests but I think we can live without them

**Comment by probablyup** [comment]
@codeaid fixed https://stackblitz.com/edit/react-ktxdtq

you can install via `styled-components@dev` for the moment until we cut a patch release in the next day or two

**Comment by codeaid** [comment]
> @codeaid fixed https://stackblitz.com/edit/react-ktxdtq

Legend! Just tested it in StackBlitz and it's indeed fixed. That's some quick turnaround!

File: src/utils/stringifyRules.js [hunk 2]
```diff
@@ -43,25 +43,63 @@ const parseRulesPlugin = _insertRulePlugin(rule => {
...
+ *
+ * & + & {
+ *   color: red;
+ * }
+ *
+ * The second ampersand should be a reference to the static component class. stylis
+ * has no knowledge of static class so we have to intelligently replace the base selector.
+ */
+const selfReferenceReplacementPlugin = (context, _, selectors) => {
+  if (context === 2 && selectors[0].lastIndexOf(_selector) > 0) {
```
Comment (toolmantim) [inline src/utils/stringifyRules.js]: We've had to revert to 4.0.0 in buildkite/site#234 because this line seems to have broken vendor prefix of webfonts? The arguments when `selectors[0]` was undefined is:

```
{ context: 2,
  selectors: [],
  _:
   'font-family:BK-Circular;src:url(/_next/static/assets/assets/fonts/lineto-circular-book-0415b07a.woff2) format("woff2");font-weight:normal;font-style:normal;font-display:swap;' }
```

I wonder if the vendor prefixing `context` of 2 is being confused with the plugin `context` of 2?

Comment (probablyup) [reply src/utils/stringifyRules.js]: If you could provide a code sample I could use in a unit test, that'd be great. In the meantime, I made this PR that should work around the specific error you mentioned: https://github.com/styled-components/styled-components/pull/2114

Comment (toolmantim) [reply src/utils/stringifyRules.js]: Thanks @probablyup 🙏🏼 That fixed it.

I've tried for an hour to create the smallest possible test case (both browser rendered and server rendered), but failed to recreate it I'm sorry. Crummy bug report I know 😔

Here's the commands to recreate it exactly though, if you needed:

```bash
git clone https://github.com/buildkite/site.git
cd site && git checkout da15e8cd
yarn && yarn start
```
